From 35f81f021c7ad545481907e2e1045544883c596a Mon Sep 17 00:00:00 2001
From: Wu Zhangjin <zhangjin.wu@windriver.com>
Date: Thu, 16 Jun 2011 15:09:50 +0800
Subject: [PATCH 02/37] nlm_xlp_64_be: add config and compile support

[ Based on netlogic SDK 20110329 ]

Adds kernel config options for XLP832 and corresponding boards

Signed-off-by: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Signed-off-by: Krishnamurthy D V <kmurthy@netlogicmicro.com>
Signed-off-by: Prasad Boddupalli <pboddupalli@netlogicmicro.com>
Integrated-by: Wu Zhangjin <zhangjin.wu@windriver.com>
---
 arch/mips/Kconfig          |   49 ++++++++++++++++++++++++++++++++++++++++++-
 arch/mips/Makefile         |   18 ++++++++++++++++
 arch/mips/netlogic/Kconfig |   15 +++++++++++++
 3 files changed, 80 insertions(+), 2 deletions(-)
 create mode 100644 arch/mips/netlogic/Kconfig

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 9c8847a..b612f3e 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -658,6 +658,33 @@ config CAVIUM_OCTEON_SIMULATOR
 	  Octeon Processor. It supports simulating Octeon processors on x86
 	  hardware.
 
+config NLM_XLP_SIM
+	bool "Support for Netlogic XLP Simulator"
+	select SMP
+	select BOOT_ELF32
+	select NLM_COMMON
+	select NLM_XLP
+	select SYS_HAS_CPU_XLP
+	select SYS_SUPPORTS_SMP
+	select HW_HAS_PCI
+	select SWAP_IO_SPACE
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_64BIT_KERNEL
+	select 64BIT_PHYS_ADDR
+	select SYS_SUPPORTS_BIG_ENDIAN
+	select SYS_SUPPORTS_LITTLE_ENDIAN
+	select SYS_SUPPORTS_HIGHMEM
+	select DMA_COHERENT
+	select CEVT_R4K
+	select CSRC_R4K
+	select IRQ_CPU
+	select ZONE_DMA if 64BIT
+	select SYNC_R4K
+	select SYS_HAS_EARLY_PRINTK
+	help
+	  This board is based on Netlogic XLP Processor.
+	  Say Y here to support this machine type
+
 config CAVIUM_OCTEON_REFERENCE_BOARD
 	bool "Cavium Networks Octeon reference board"
 	select CEVT_R4K
@@ -697,6 +724,7 @@ source "arch/mips/sgi-ip27/Kconfig"
 source "arch/mips/sibyte/Kconfig"
 source "arch/mips/txx9/Kconfig"
 source "arch/mips/vr41xx/Kconfig"
+source "arch/mips/netlogic/Kconfig"
 source "arch/mips/cavium-octeon/Kconfig"
 source "arch/mips/loongson/Kconfig"
 
@@ -1291,6 +1319,19 @@ config CPU_SB1
 	select CPU_SUPPORTS_HIGHMEM
 	select WEAK_ORDERING
 
+config CPU_XLP
+	bool "XLP"
+	depends on SYS_HAS_CPU_XLP
+	select CPU_SUPPORTS_32BIT_KERNEL
+	select CPU_SUPPORTS_64BIT_KERNEL
+	select CPU_SUPPORTS_HIGHMEM
+	select CPU_HAS_LLSC
+	select WEAK_ORDERING
+	select WEAK_REORDERING_BEYOND_LLSC
+	select CPU_HAS_PREFETCH
+	help
+	  Netlogic Corporation XLP processors.
+
 config CPU_CAVIUM_OCTEON
 	bool "Cavium Octeon processor"
 	depends on SYS_HAS_CPU_CAVIUM_OCTEON
@@ -1440,6 +1481,10 @@ config WEAK_ORDERING
 #
 config WEAK_REORDERING_BEYOND_LLSC
 	bool
+
+config SYS_HAS_CPU_XLP
+	bool
+
 endmenu
 
 #
@@ -1462,7 +1507,7 @@ config CPU_MIPSR1
 
 config CPU_MIPSR2
 	bool
-	default y if CPU_MIPS32_R2 || CPU_MIPS64_R2 || CPU_CAVIUM_OCTEON
+	default y if CPU_MIPS32_R2 || CPU_MIPS64_R2 || CPU_CAVIUM_OCTEON || CPU_XLP
 
 config SYS_SUPPORTS_32BIT_KERNEL
 	bool
@@ -1482,7 +1527,7 @@ config CPU_SUPPORTS_UNCACHED_ACCELERATED
 	bool
 config MIPS_PGD_C0_CONTEXT
 	bool
-	default y if 64BIT && CPU_MIPSR2
+	default y if 64BIT && CPU_MIPSR2 && !CPU_XLP
 
 #
 # Set to y for ptrace access to watch registers.
diff --git a/arch/mips/Makefile b/arch/mips/Makefile
index 2e4fe44..8ef0af4 100644
--- a/arch/mips/Makefile
+++ b/arch/mips/Makefile
@@ -179,6 +179,7 @@ ifeq (,$(findstring march=octeon, $(cflags-$(CONFIG_CPU_CAVIUM_OCTEON))))
 cflags-$(CONFIG_CPU_CAVIUM_OCTEON) += -Wa,-march=octeon
 endif
 
+cflags-$(CONFIG_CPU_XLP)    += $(call cc-option,-march=xlp) -Wa,--trap
 cflags-$(CONFIG_CPU_R4000_WORKAROUNDS)	+= $(call cc-option,-mfix-r4000,)
 cflags-$(CONFIG_CPU_R4400_WORKAROUNDS)	+= $(call cc-option,-mfix-r4400,)
 cflags-$(CONFIG_CPU_DADDI_WORKAROUNDS)	+= $(call cc-option,-mno-daddi,)
@@ -648,6 +649,23 @@ core-$(CONFIG_TOSHIBA_RBTX4938) += arch/mips/txx9/rbtx4938/
 core-$(CONFIG_TOSHIBA_RBTX4939) += arch/mips/txx9/rbtx4939/
 
 #
+# NETLOGIC SOC Common (common)
+core-$(CONFIG_NLM_COMMON)	+= arch/mips/netlogic/boot/
+core-$(CONFIG_NLM_COMMON)	+= arch/mips/netlogic/common/
+cflags-$(CONFIG_NLM_COMMON)	+= -I$(srctree)/arch/mips/include/asm/mach-netlogic
+cflags-$(CONFIG_NLM_COMMON)	+= -I$(objtree)/arch/mips/netlogic/boot
+cflags-$(CONFIG_NLM_COMMON)	+= -I$(srctree)/arch/mips/netlogic/boot
+cflags-$(CONFIG_NLM_COMMON)	+= -I$(srctree)/arch/mips/include/asm/netlogic
+
+#
+# NETLOGIC XLP Soc, Simulator and boards
+#
+core-$(CONFIG_NLM_XLP)		+= arch/mips/netlogic/xlp/
+cflags-$(CONFIG_NLM_XLP)	+= -I$(srctree)/arch/mips/include/asm/netlogic/hal
+# This address is now configured via kernel configuration file
+load-$(CONFIG_NLM_XLP_SIM)	+= $(CONFIG_NLM_COMMON_LOAD_ADDRESS)
+
+#
 # Cavium Octeon
 #
 core-$(CONFIG_CPU_CAVIUM_OCTEON)	+= arch/mips/cavium-octeon/
diff --git a/arch/mips/netlogic/Kconfig b/arch/mips/netlogic/Kconfig
new file mode 100644
index 0000000..1c28804
--- /dev/null
+++ b/arch/mips/netlogic/Kconfig
@@ -0,0 +1,15 @@
+config NLM_COMMON
+	bool
+
+config NLM_XLP
+	bool
+
+config NLM_COMMON_LOAD_ADDRESS
+	hex "Netlogic Linux kernel start address"
+	depends on NLM_COMMON
+	default "0xffffffff84000000"
+	help
+	  This is start address for the linux kernel. Default value
+          should be good for most of the applications unless specified
+          explicitly: e.g. running Netlogic ToE requires kernel to be linked
+	  at address 0xffffffff86000000.
-- 
1.7.0.2

